perf: faster Payroll  :- Part 1 (Payroll Enty)

saurabh6790 opened the pull request:
Tested on 100k Employee records

## Employee Selections
#### Before:
<img width="1434" alt="Screenshot 2023-04-03 at 4 31 41 PM" src="https://user-images.githubusercontent.com/3784093/229491930-ad9bc14c-4fca-4bb0-82d6-20817483c15c.png">

search_link: apx 2.23 sec

#### After:

<img width="1440" alt="Screenshot 2023-04-03 at 4 33 45 PM" src="https://user-images.githubusercontent.com/3784093/229492227-10891f69-ad3e-4141-b5e1-dd8e81bd17b6.png">

search_link: apx 62 ms

---
Data Chunk 7k employees

## Fill Employee table based on filters
### Before
<img width="1432" alt="Screenshot 2023-04-03 at 4 32 43 PM" src="https://user-images.githubusercontent.com/3784093/229491957-0fae2f3e-8f4e-4fb7-9ff2-25999122fff0.png">

run_doc_method = 12.14
savedoc = 19.29

### After
<img width="1433" alt="Screenshot 2023-04-03 at 4 34 48 PM" src="https://user-images.githubusercontent.com/3784093/229492384-06283d7d-a138-4dad-b468-7abb8990ab37.png">


run_doc_method = 10.16 sec
savedoc = 19.37 sec

---

## validate salary slip against employees
### Before

<img width="701" alt="Screenshot 2023-04-10 at 3 29 14 PM" src="https://user-images.githubusercontent.com/3784093/230880350-e8e16ef9-94f5-4e17-ab0c-9659f90b9507.png">

Time Required = 7 Sec

### After
<img width="764" alt="Screenshot 2023-04-10 at 3 29 41 PM" src="https://user-images.githubusercontent.com/3784093/230880397-9b9af4d9-4967-47ce-8fb4-692fcae47e79.png">

Time Required = 0.166

---
## Validate employee attendance

### Before
<img width="728" alt="Screenshot 2023-04-10 at 5 24 55 PM" src="https://user-images.githubusercontent.com/3784093/230899293-2e84faec-711e-41f9-9367-777f890318bb.png">

### After
<img width="806" alt="Screenshot 2023-04-10 at 5 40 22 PM" src="https://user-images.githubusercontent.com/3784093/230899323-7d6313f5-ec36-48fd-a701-0e449a95a960.png">

---
## make_payment_entry

### Before
<img width="630" alt="Screenshot 2023-04-12 at 12 54 11 PM" src="https://user-images.githubusercontent.com/3784093/231384396-a436decb-fb43-40bc-aca1-add7959555d5.png">

### After
<img width="700" alt="Screenshot 2023-04-12 at 12 55 40 PM" src="https://user-images.githubusercontent.com/3784093/231384444-e21b3a54-6e39-448b-94e0-f7c67665088b.png">

saurabh6790 committed:
feat: remove ifnull check from conditions to consider indexing

saurabh6790 committed:
feat: remove duplicate code

saurabh6790 committed:
chore: code cleanup

saurabh6790 committed:
chore: code cleanup

saurabh6790 committed:
fix: validate existing salary slips via single query

saurabh6790 committed:
fix: validate_employee_attendance

saurabh6790 committed:
fix: handle cancellation via background job for higher employees count

saurabh6790 committed:
fix: make_payment_entry

saurabh6790 committed:
fix: optimise make_payment_entry

saurabh6790 committed:
fix: run_doc_method

saurabh6790 committed:
chore: reorganise code and test fix

saurabh6790 committed:
chore: merge common code and break larger functions into smaller functions

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -105,24 +105,26 @@ frappe.ui.form.on('Payroll Entry', {
 			doc: frm.doc,
 			method: 'fill_employee_details',
+			freeze: true,
+			freeze_message: __('Pulling Employees')
```
```suggestion
			freeze_message: __('Fetching Employees')
```
fetching sounds better than pulling?

ruchamahabal replied:
```suggestion
			freeze_message: __('Fetching Employees')
```
Fetching sounds better than pulling?

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -105,24 +105,26 @@ frappe.ui.form.on('Payroll Entry', {
-				}
-				frm.scroll_to_field("employees");
+			frm.refresh();
+			if (r.docs[0].validate_attendance) {
```
```suggestion
			if (r.docs?.[0]?.validate_attendance) {
```
Optional chaining to make sure value exists

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -105,24 +105,26 @@ frappe.ui.form.on('Payroll Entry', {
 		frm.call({
 			doc: frm.doc,
-			method: "create_salary_slips",
+			method: "run_doc_method",
+			args: {
+				method: "create_salary_slips",
+				dt: "Payroll Entry",
+				dn: frm.doc.name
+			},
```
Why do we need to use `run_doc_method`? frm.call with the doc will directly call the `create_salary_slips` anyway. `run_doc_method` usage is not very standard across the frappe codebase.

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.js:
```diff
@@ -389,8 +391,12 @@ let make_bank_entry = function (frm) {
 		return frappe.call({
-			doc: cur_frm.doc,
-			method: "make_payment_entry",
+			method: "run_doc_method",
+			args: {
+				method: "make_bank_entry",
+				dt: "Payroll Entry",
+				dn: frm.doc.name
+			},
```
Again, why `run_doc_method`

saurabh6790 replied:
If we pass frm.doc as argument then frm.call internally calls `run_doc_method` to run a class method. 

While testing over a large data of 7k employees, the system passing whole 3.2 mb object from client to server.
But after submitting the payroll entry, there will no change in data, so we do not have to pass whole doc from client side to server. Thus used `run_doc_method` explicitly

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -64,19 +53,31 @@ def set_status(self, status=None, update=False):
+	def before_submit(self):
+		self.validate_employee_details()
+		self.validate_payroll_payable_account()
+		if self.validate_attendance:
```
Let's move this check into `validate_employee_attendance` method itself with an early return.
It's being checked again at line number 168

https://github.com/frappe/hrms/pull/417/files#diff-9ab3d3f35681d26efdd18a07bd2422c6e6833df240dc01f39c80450821b1f8f5R168

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -756,138 +839,228 @@ def set_start_end_dates(self):
 
 	@frappe.whitelist()
 	def validate_employee_attendance(self):
+		"""{
```
Mention what the method returns before specifying the format

`"""returns a dict of employee attendance details`

We should actually add type hints to these methods

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -756,138 +839,228 @@ def set_start_end_dates(self):
+			if not employee_details[employee]["holiday_list"]:
+				employee_details[employee]["holiday_list"] = default_holiday_list
+
+		for emp in self.employees:
```
Wondering if this can be simplified. We are iterating over the same list thrice - list comprehension, adding attendance count and holiday list and then the final loop which checks everything.

saurabh6790 replied:
https://github.com/frappe/hrms/pull/417/commits/396c5f0911b33f1a1ade279fd3cd4a5105bcf5d9

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -64,19 +53,31 @@ def set_status(self, status=None, update=False):
+		self.validate_payroll_payable_account()
+		if self.validate_attendance:
+			if self.validate_employee_attendance():
+				frappe.throw(_("Cannot Submit, Employees left to mark attendance"))
```
```suggestion
				frappe.throw(_("Cannot Submit. Attendance is not marked for some employees."))
```
Also this statement should move to `validate_employee_attendance`

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -64,19 +53,31 @@ def set_status(self, status=None, update=False):
+		self.set_status(update=True, status="Submitted")
+		self.create_salary_slips()
+
 	def validate_employee_details(self):
```
```suggestion
	def validate_existing_salary_slips(self):
```
Let's rename this to what it's actually doing?

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -756,138 +839,228 @@ def set_start_end_dates(self):
 
+	return query.run(as_dict=as_dict)
 
-def get_joining_relieving_condition(start_date, end_date):
```
This condition is not being checked in `get_employee_list` at all now?

saurabh6790 replied:
https://github.com/frappe/hrms/pull/417/files#diff-9ab3d3f35681d26efdd18a07bd2422c6e6833df240dc01f39c80450821b1f8f5R980-R981

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -112,39 +116,32 @@ def get_emp_list(self):
+			branch=self.branch,
+			department=self.department,
+			designation=self.designation,
+			currency=self.currency,
+			payroll_frequency=self.payroll_frequency,
+			start_date=self.start_date,
+			end_date=self.end_date,
+			payroll_payable_account=self.payroll_payable_account,
+			salary_slip_based_on_timesheet=self.salary_slip_based_on_timesheet,
```
These columns are not present in employee master. Why add these filters here?

saurabh6790 replied:
Same filters dict we are using for multiple purposes.
- to get salary structure
- to get employees based on branch, company, department, designation. Here start date and end date is used to validate against joining date and reliving date.

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -112,39 +116,32 @@ def get_emp_list(self):
+	def check_mandatory(self):
+		for fieldname in ["company", "start_date", "end_date"]:
+			if not self.get(fieldname):
+				frappe.throw(_("Please set {0}").format(self.meta.get_label(fieldname)))
```
I don't think we need this. These fields are mandatory so shouldn't need a separate validation

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -406,87 +403,160 @@ def make_accrual_jv_entry(self):
+				self.update_salary_slip_status(submitted_salary_slips, jv_name=journal_entry.name)
+
+		except Exception as e:
+			frappe.log_error(
+				frappe.get_traceback(), reference_doctype="Payroll Entry", reference_name=self.name
+			)
```
```suggestion
			self.log_error("Journal Entry creation against Salary Slip failed")
```
This will log everything with doc references

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -592,84 +658,105 @@ def get_amount_and_exchange_rate_for_journal_entry(
 
 	@frappe.whitelist()
-	def make_payment_entry(self):
+	def make_bank_entry(self):
```
Renaming a whitelisted method is a breaking change. Since this will be shipped to minor versions too, I don't think we should be doing it here. People might have dependent code on this.

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -1082,16 +1255,20 @@ def show_payroll_submission_status(submitted, unsubmitted, payroll_entry):
+
+	return (
+		frappe.qb.from_(SalarySlip)
+		.select(SalarySlip.employee)
```
```suggestion
		.select(SalarySlip.employee)
		.distinct()
```
Missed distinct from the above query

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -1024,21 +1197,21 @@ def log_payroll_failure(process, payroll_entry, error):
+			count += 1
+			if publish_progress:
+				frappe.publish_progress(
+					count * 100 / len(set(employees) - set(salary_slips_exist_for)),
```
```suggestion
					count * 100 / len(employees),
```

Can directly use employees here since duplicates are already removed on line number 1204

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -756,138 +839,228 @@ def set_start_end_dates(self):
+	).run(pluck=True)
+
+
+def get_emp_list(
```
This whole get_emp_list function has become extremely confusing now. There is a method called `get_emp_list`, a function `get_employee_list` and one more function `get_emp_list` outside the class.

Commonifying code is okay until it affects readability. This particular scenario is making it more difficult to figure out what's happening. Rethink this whole refactor maybe.

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -406,87 +403,160 @@ def make_accrual_jv_entry(self):
-					entry_type="credit",
-				)
-				accounts.append(accounting_entry)
+			self.set_payable_amuont_against_payroll_payable_account(
```
```suggestion
			self.set_payable_amount_against_payroll_payable_account(
```

Spellcheck PRs

ruchamahabal commented:
Better commit conventions can be used:

`perf:` for performance fixes
`refactor:` for such commits instead of `chore:`: [`04a1908` (#417)](https://github.com/frappe/hrms/pull/417/commits/04a190841174382b590cd9f65cf7b302d981ca8d)

saurabh6790 committed:
chore: fix method naming and empty state check

saurabh6790 committed:
chore: fix test case

saurabh6790 committed:
fix: test case

saurabh6790 committed:
fix: validation for sending email against salary slip in test

saurabh6790 committed:
fix: error loggin and remove unwanted methods

saurabh6790 committed:
feat: optimize method to validate employees to mark attendance

saurabh6790 committed:
fix: validate employee attendance

saurabh6790 committed:
fix: test case

ruchamahabal committed:
fix: rearrange email ss checks to save db calls

- remove unnecessary log error params

- fix attendance message

ruchamahabal committed:
revert: "fix: handle cancellation via background job for higher employees count"

This reverts commit cdb48c9fa66f98040cc0eec192657a841bd003f3.

ruchamahabal commented:
reverting "fix: handle cancellation via background job for higher employees count" 

Reasons:
- Queuing salary slip deletion needs more handling and UX work. The current one doesn’t function correctly: status changes to cancelled immediately but it shows as queued
- Also if salary slip is submitted, deletion fails. ignoring gl entries during deletion is also not working for some reason.

https://user-images.githubusercontent.com/24353136/237057318-798d91be-e3ce-44e7-9ad1-0918ceb31ea9.mp4

Will be taken care of in a separate PR

ruchamahabal committed:
refactor: multiple fixes

- `build_query` renamed to `get_filtered_employees` for clarity

- rectified query condition for ssa date

- fix syntax not_in -> notin

- fix type emplyees -> employees

- fix: apply limit & offset separately

ruchamahabal committed:
refactor: query fixes

- rename variable `process_payroll_accounting_entry_based_on_employee` to `employee_wise_accounting_enabled` for better readability

- add missing salary structure where clause in payroll cost center query

- fix incorrect function docstrings and add type hints

- add missing distinct clause while fetching employee fields

- fix pluck usage in `get_existing_salary_slips`

ruchamahabal commented on hrms/payroll/doctype/payroll_entry/payroll_entry.py:
```diff
@@ -610,71 +667,89 @@ def get_amount_and_exchange_rate_for_journal_entry(
-					)
-					if only_tax_impact != 1 and statistical_component != 1:
-						if is_flexible_benefit == 1 and creat_separate_je == 1:
-							self.create_journal_entry(sal_detail.amount, sal_detail.salary_component)
```
This separate je against benefit claim is not getting created in the new refactor. Missed adding this?

saurabh6790 replied:
`set_accounting_entries_for_bank_entry` handles it

ruchamahabal committed:
fix(typo): `creat_separate_je` -> `create_separate_je`

ruchamahabal committed:
Merge branch 'develop' into profiling-fixes
